Sources/MapboxNavigation/NavigationMapView.swift

```diff
@@ -374,18 +409,60 @@ open class NavigationMapView: UIView {
+                guard let screenCoordinate = self?.mapView.screenCoordinate(for: location.coordinate) else { return }
+                self?.userCourseView.center = CGPoint(x: screenCoordinate.x, y: screenCoordinate.y)
+            }
+            
+            let cameraOptions = CameraOptions(cameraState: mapView.cameraState)
+            userCourseView.update(location: location,
+                                  pitch: cameraOptions.pitch!,
+                                  direction: cameraOptions.bearing!,
+                                  animated: animated,
+                                  navigationCameraState: navigationCamera.state)
+        case .puck2D(configuration: _):
+            if simulation {
+                if !passiveLocationUsed {
+                    let locationProvider = mapView.location.locationProvider
+                    mapView.location.locationProvider(locationProvider!, didUpdateLocations: [location])
+                    mapView.location.locationProvider.stopUpdatingLocation()
+                }
+                if mapView.mapboxMap.style.layerExists(withId: NavigationMapView.LayerIdentifier.puck2DLayer) {
+                    do {
+                        try mapView.mapboxMap.style.setLayerProperty(for: NavigationMapView.LayerIdentifier.puck2DLayer,
+                                                                     property: "bearing",
+                                                                     value: location.course)
+                    } catch {
+                        NSLog("Failed to perform operation while updating 2D puck bearing arrow with error: \(error.localizedDescription).")
+                    }
+                }
+            }
+        case .puck3D(configuration: let configuration):
+            if simulation {
+                if !passiveLocationUsed {
+                    mapView.location.locationProvider.stopUpdatingLocation()
+                }
+                if mapView.mapboxMap.style.sourceExists(withId: NavigationMapView.SourceIdentifier.puck3DSource) {
+                    var model = configuration.model
+                    model.position = [location.coordinate.longitude, location.coordinate.latitude]
+                    if var orientation = model.orientation,
+                       orientation.count > 2 {
+                        orientation[2] = orientation[2] + location.course
+                        model.orientation = orientation
+                    }
+                    
+                    let modelSourceCode = [NavigationMapView.ModelKeyIdentifier.modelSouce: model]
+                    if let data = try? JSONEncoder().encode(modelSourceCode),
+                       let jsonDictionary = try? JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] {
+                        try? mapView.mapboxMap.style.setSourceProperty(for: NavigationMapView.SourceIdentifier.puck3DSource, property: "models", value: jsonDictionary)
```

Comment:
Is it possible to use Model more directly, rather than serializing to JSON?

Reply:
Because the source type of the 3D puck layer is`ModelSource`, which is internal and could not be updated directly.

Reply:
ModelSource became internal in mapbox/mapbox-maps-ios#198. I don’t see any rationale associated with it, so it might have been just for lack of external demand. ModelSource would be more type-safe and forward-compatible than raw JSON usage. Since this is how the map SDK currently requires you to use the `models` property, let’s track ModelSource usage as tail work and also request an upstream change.

/cc @macdrevx